src/kernel/network.cpp

```diff
@@ -482,37 +579,37 @@ void CryptoKernel::Network::connectionFunc() {
             log->printf(LOG_LEVEL_INFO,
                         "Network(): Peer connected from " + client->getRemoteAddress().toString() + ":" +
                         std::to_string(client->getRemotePort()));
-            PeerInfo* peerInfo = new PeerInfo();
-            peerInfo->peer.reset(new Peer(client, blockchain, this, true));
+            Connection* connection = new Connection(); // todo check for thread safety (think it's okay)
+            connection->setPeer(new Peer(client, blockchain, this, true));
 
             Json::Value info;
 
             try {
-                info = peerInfo->peer->getInfo();
+                info = connection->getInfo();
             } catch(const Peer::NetworkError& e) {
                 log->printf(LOG_LEVEL_WARN, "Network(): Failed to get information from connecting peer");
-                delete peerInfo;
+                delete connection;
                 continue;
             }
 
             try {
-                peerInfo->info["height"] = info["tipHeight"].asUInt64();
-                peerInfo->info["version"] = info["version"].asString();
+                connection->setInfo("height", info["tipHeight"].asUInt64());
+                connection->setInfo("version", info["version"].asString());
             } catch(const Json::Exception& e) {
                 log->printf(LOG_LEVEL_WARN, "Network(): Incoming peer sent invalid info message");
-                delete peerInfo;
+                delete connection;
                 continue;
             }
 
             const std::time_t result = std::time(nullptr);
-            peerInfo->info["lastseen"] = static_cast<uint64_t>(result);
+            connection->setInfo("lastseen", static_cast<uint64_t>(result));
 
-            peerInfo->info["score"] = 0;
+            connection->setInfo("score", 0);
 
-            connected[client->getRemoteAddress().toString()].reset(peerInfo);
+            connected.at(client->getRemoteAddress().toString()).reset(connection);
 
             std::unique_ptr<Storage::Transaction> dbTx(networkdb->begin());
-            peers->put(dbTx.get(), client->getRemoteAddress().toString(), peerInfo->info);
+            peers->put(dbTx.get(), client->getRemoteAddress().toString(), connection->getInfo());
```

Comment:
This needs to use the cached info value inside connection or there could be a `Peer::NetworkError` exception. It's also more efficient to use the cached version since it was just set above.